### PR TITLE
vmm: refactor vcpu.rs file

### PR DESCRIPTION
**Pull request by dianpopa**
Break up vcpu into 3 files: one that has
common functionality for both architectures (mainly the
one in charge of signal handling) and 2 files per architecture.

Signed-off-by: Diana Popa <dpopa@amazon.com>

## Reason for This PR

The vstate/vcpu.rs file used to contain all functionality related to interacting with the vcpus: signaling, handling events, x86_64 functionality and aarch64 functionality. This made the file really big (i.e aproximately 1500 lines).
This PR breaks the vcpu.rs into 3 files:
* mod.rs which contains functionality common to all architectures (signaling, events)
* x86_64: kvm specific interaction which is x86_64 specific
* aarch64: kvm specific interaction which is aarch64 specific

The vcpu structure now contains a new structure in it (kvm_vcpu) which contains the kvm and architecture specific functionality.

## Description of Changes

`[Author TODO: add description of changes.]`

- [ ] This functionality can be added in [`rust-vmm`](https://github.com/rust-vmm).

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The reason for this PR is clearly provided (issue no. or explanation).
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] Any newly added `unsafe` code is properly documented.
- [x] Any API changes are reflected in `firecracker/swagger.yaml`.
- [x] Any user-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.

### CODE REVIEW

File: src/vmm/src/vstate/vm.rs
Comment (serban300) [inline src/vmm/src/vstate/vm.rs:312]: Nit: is this comment needed ?

**Review by serban300 — APPROVED** [review]

File: src/vmm/src/vstate/vm.rs
Comment (dianpopa) [reply src/vmm/src/vstate/vm.rs:312]: No, no, I'll delete it when I do the rebase and get a second review.

**Review by iulianbarbu — APPROVED** [review]

File: src/vmm/src/vstate/vcpu/x86_64.rs
Comment (georgepisaltu) [inline src/vmm/src/vstate/vcpu/x86_64.rs:357]: nit: I would rename `r` to something like `unexpected_exit`.

File: src/vmm/src/vstate/vcpu/mod.rs
Comment (georgepisaltu) [inline src/vmm/src/vstate/vcpu/mod.rs:480]: nit: Maybe rename `r` to `uncommon_reason` or `arch_specific_reason`.

File: src/vmm/src/vstate/vcpu/mod.rs
Comment (georgepisaltu) [inline src/vmm/src/vstate/vcpu/mod.rs:222]: Is there a way to move this to the x86 specific file?

File: src/vmm/src/vstate/vcpu/mod.rs
Comment (georgepisaltu) [inline src/vmm/src/vstate/vcpu/mod.rs:482]: nit: Rename this to `run_arch_emulation()` for clarity.

File: src/vmm/src/vstate/vm.rs
Comment (dianpopa) [reply src/vmm/src/vstate/vm.rs]: Done, take a look.

File: src/vmm/src/vstate/vcpu/mod.rs
Comment (dianpopa) [reply src/vmm/src/vstate/vcpu/mod.rs]: Indeed, good idea, moved it. Take a look!

File: src/vmm/src/vstate/vcpu/mod.rs
Comment (dianpopa) [reply src/vmm/src/vstate/vcpu/mod.rs]: Renamed.

File: src/vmm/src/vstate/vcpu/mod.rs
Comment (dianpopa) [reply src/vmm/src/vstate/vcpu/mod.rs]: Yup, renamed.

File: src/vmm/src/vstate/vcpu/x86_64.rs
Comment (dianpopa) [reply src/vmm/src/vstate/vcpu/x86_64.rs]: Renamed.

**Review by dianpopa — COMMENTED** [review]
@serban300 , @iulianbarbu , @georgepisaltu I addressed your comments. Please take another look!

**Review by serban300 — APPROVED** [review]

File: src/vmm/src/vstate/vcpu/aarch64.rs
Comment (acatangiu) [inline src/vmm/src/vstate/vcpu/aarch64.rs:57]: nit:
```suggestion
/// A wrapper around creating and using a kvm aarch64 vcpu.
```

File: src/vmm/src/lib.rs
Comment (acatangiu) [inline src/vmm/src/lib.rs:250]: nit: why the curly braces?

File: src/vmm/src/vstate/vcpu/aarch64.rs
Comment (acatangiu) [inline src/vmm/src/vstate/vcpu/aarch64.rs:74]: nit: comment refers old params

File: src/vmm/src/vstate/vcpu/mod.rs
Comment (acatangiu) [inline src/vmm/src/vstate/vcpu/mod.rs:100]: nit:
this is `pub`, everything in `KvmVcpu` is also `pub`, I prefer we didn't allow mutable access to `kvm_vcpu` from outside the `Vcpu`.
It'd be less error-prone if we did like in the case of `mmio_bus` where we have `Vcpu::set_mmio_bus()` that internally sets `vcpu.kvm_vcpu.mmio_bus`

File: src/vmm/src/vstate/vcpu/x86_64.rs
Comment (acatangiu) [inline src/vmm/src/vstate/vcpu/x86_64.rs:147]: nit: this is only used to configure for boot, the `configure` function is also arch specific so it could simply take this as a param to be used locally. That would avoid unnecessarily carrying this around in memory.

File: src/vmm/src/vstate/vcpu/x86_64.rs
Comment (acatangiu) [inline src/vmm/src/vstate/vcpu/x86_64.rs:155]: nit: doc comment outdated

File: src/vmm/src/vstate/vcpu/x86_64.rs
Comment (acatangiu) [inline src/vmm/src/vstate/vcpu/x86_64.rs:264]: nit: unrelated to this PR, but UH-OH! This leftover `assert` here looks dangerous, it should be a `Error::VcpuGetMsrs` instead of an induced crash.

(the assert around msr indices is fine as that validates code implementation logic, but this one depends on the value returned from `kvm` and can actually trigger with no fault in code logic)

File: src/vmm/src/vstate/vcpu/x86_64.rs
Comment (acatangiu) [inline src/vmm/src/vstate/vcpu/x86_64.rs:386]: nit: I think this is only used in tests, maybe move it there.

**Review by serban300 — APPROVED** [review]

**Pull request merged**
